run-test.sh

```diff
@@ -11,55 +11,110 @@ set_debug
+    )
 }
 
-run() {
-    create_test_files
+run_test() {
+    cleanup "${ENCRYPTED_PATH}"
+
+    # Poorman's synchronisation; giving test-data-generator service (create_test_files) a chance to complete.
+    sleep 2
 
     # perform backup
     /app/entrypoint.sh backup
 
-    # if s3sync is enabled, then remove local encrypted files so that they are pulled down from S3, during restore
-    # attempts below
-    if s3sync is_enabled; then
-        rm -f "${ENCRYPTED_PATH}/"*.txt.gpg
-    fi
+    # Verify that SOURCE_REMOVE_PLAIN option takes effect
+    for i in $(seq 1 "$n_files"); do
+        local filepath="${SOURCE_PATH}/data-$(printf '%02d' "$i").txt"
+        if [ "$SOURCE_REMOVE_PLAIN" = 1 ]; then
+            bail_file_exists "$filepath"
+        else
+            bail_file_not_exists "$filepath"
+        fi
+    done
 
-    # restore each file
-    for i in $(seq 1 5); do
-        local filepath="${SOURCE_PATH}/data-$(printf %02d $i).txt"
-        local filename=$(basename "$filepath")
+    # Restore each file only if the SOURCE_PATH is writable.
+    # SOURCE_PATH can be made read-only via settings given to volume options
+    # This test makes sure that gpg-s3sync honours these and does not fall apart.
+    if [ -w "${SOURCE_PATH}" ]; then
+        for i in $(seq 1 "$n_files"); do
+            local filepath="${SOURCE_PATH}/data-$(printf '%02d' "$i").txt"
+            local filename=$(basename "$filepath")
 
-        if is_number "$SOURCE_REMOVE_PLAIN" && [ "$SOURCE_REMOVE_PLAIN" -eq 1 ]; then
-            bail_file_exists "$filepath";
-        fi
+            # remove the file before restore; otherwise 'restore' will bail seeing the file
+            rm -f "$filepath";
 
-        /app/entrypoint.sh restore "${filename}.gpg"
-        verify_test_file "$filename"
-    done
+            # if s3sync is enabled, then remove local encrypted files so that they are pulled down from S3, during
+            # restore attempts below
+            if s3sync is_enabled; then
+                rm -f "${ENCRYPTED_PATH}/"*.txt.gpg
+            fi
+
+            # restore the specific file
+            /app/entrypoint.sh restore "${filename}.gpg"
+            verify_test_file "$filename"
+        done
 
-    # Using the last file from the above set of tests, which has already been restored into SOURCE_PATH, check if the
-    # restore will fail if the destination file already exists. If the command succeeds, i.e the file was overwritten,
-    # then it is treated as a failure.
-    if /app/entrypoint.sh restore "data-01.txt.gpg"; then
-        error "File in SOURCE_PATH overwritten!"
-        return 1;
+        # Using one file from the above set of tests, which has already been restored into SOURCE_PATH, check if the
+        # restore will fail if the destination file already exists. If the command succeeds, i.e the file was
+        # overwritten, then it is treated as a failure.
+        if /app/entrypoint.sh restore "data-01.txt.gpg"; then
```

Comment:
Potentially this could be done before the clean up (line 84)?  Which is maybe a little more representative, as the file will have been created by the application rather than the syncer.